pkg/epp/datalayer/metrics/client.go

```diff
@@ -0,0 +1,135 @@
+/*
+Copyright 2025 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package metrics
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"net/url"
+	"sync"
+	"time"
+
+	"github.com/prometheus/common/expfmt"
+	"sigs.k8s.io/gateway-api-inference-extension/pkg/epp/datalayer"
+)
+
+// Client is an interface for retrieving the metrics from an endpoint URL.
+type Client interface {
+	Get(ctx context.Context, target *url.URL, ep datalayer.Addressable) (PrometheusMetricMap, error)
+}
+
+// ClientFactory returns a Client suitable for an endpoint.
+// Implementations may return a new Client each time, or use a cached
+// copy for optimized retrieval.
+type ClientFactory interface {
+	GetClientForEndpoint(ep datalayer.Addressable) Client
+}
+
+// GetDefaultClientFactory returns a default implementation of the
+// ClientFactory, which caches and reuses client across calls.
+func GetDefaultClientFactory() ClientFactory {
+	return defaultClientFactory
+}
+
+// -- package implementations --
+var (
+	cleanupTick          = 30 * time.Second
+	maxIdleTime          = time.Minute
+	defaultClientFactory = newClientFactory()
+)
+
+type client struct {
+	cl       *http.Client
+	lastUsed time.Time
+}
+
+type clientmap struct {
+	cleanupOnce sync.Once
+	clients     sync.Map // map[string]*client (uses client IP as key)
+}
+
+func newClientFactory() *clientmap {
+	clm := &clientmap{}
+	clm.startCleanupGoroutine()
```

Comment:
nit: do we really need the cleanupOnce?
can't we do just:
```go
clm := &clientmap{}
go clm.startCleanup() // (without cleanupOnce, without go func() inside)
...
```

Reply:
it goes back to the use of newClientFactory. If we can always guarantee it'll be called once, then - yes, the suggestion works.
The use of the var above and the once here was to guard against multiple calls to GetDefaultClientFactory.